portfolio/src/main/java/com/google/sps/data/SentimentAnalyzer.java

```diff
@@ -0,0 +1,163 @@
+import com.google.cloud.language.v1.ClassificationCategory;
+import com.google.cloud.language.v1.Document;
+import com.google.cloud.language.v1.Document.Type;
+import com.google.cloud.language.v1.Entity;
+import com.google.cloud.language.v1.LanguageServiceClient;
+import com.google.cloud.language.v1.Sentiment;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSet.Builder;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Uses the Natural Language API to perform entity and sentiment analysis.
+ */
+public class SentimentAnalyzer {
+  private static LanguageServiceClient languageService;
+  private static final String SALIENCE_PROPERTY = "SALIENCE";
+  private static final String SENTIMENT_PROPERTY = "SENTIMENT";
+
+  /**
+   * Constructor that instantiates the LanguageServiceClient
+   */
+  public SentimentAnalyzer() throws IOException {
+    languageService = LanguageServiceClient.create();
+  }
+
+  /**
+   * Checks if the given text has a positive sentiment
+   *
+   * @param text: Text to be analyzed
+   */
+  public static boolean hasPositiveSentiment(String text) throws IOException {
+    return getSentimentScore(text) > 0;
+  }
+
+  /**
+   * Gets sentiment score (between -1 and 1) for given text
+   *
+   * @param text: Text to be analyzed
+   * @return float score, where -1 is the negative extreme and +1 is the positive extreme
+   */
+  public static float getSentimentScore(String text) throws IOException {
+    Document doc = makeDocument(text);
+    Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
+    float score = sentiment.getScore();
+    return score;
+  }
+
+  /**
+   * Gets categories and their associated confidences for given text
+   * Complete list of categories: cloud.google.com/natural-language/docs/categories
+   *
+   * @param text: Text to be classified
+   * @return ImmutableMap from categories (as Strings) to confidences (as floats)
+   */
+  public ImmutableMap<String, Float> getCategories(String text) {
+    ImmutableMap.Builder<String, Float> strCategories = new ImmutableMap.Builder<String, Float>();
+    Document doc = makeDocument(text);
+    List<ClassificationCategory> categories = languageService.classifyText(doc).getCategoriesList();
+    for (ClassificationCategory category : categories) {
+      strCategories.put(category.getName(), category.getConfidence());
+    }
+    return strCategories.build();
+  }
+
+  /**
+   * Gets an immutableSet of entities in given text
+   *
+   * @param text: Text to be analyzed
+   * @return ImmutableSet of entities found, represented as Strings
+   */
+  public ImmutableSet<String> getEntities(String text) {
+    return getEntitiesWithSentiments(text).keySet();
```

Comment:
You could also alter this method so that you can use it as a helper function in getEntitiesWith and getEntitiesWithType. It seems like it's the same code in both of them so I think it would be a good idea.